File: Objc/OCTCustomCollectionViewLayout/OCTCustomCollectionViewLayout/OCTMainViewController.m
//
//  ViewController.m
//  OCTCustomCollectionViewLayout
//
//  Created by dmitry.brovkin on 4/3/17.
//  Copyright © 2017 dmitry.brovkin. All rights reserved.
//

#import "OCTMainViewController.h"

static NSString *const kReuseCellID = @"reuseCellID";


@interface OCTMainViewController () <UICollectionViewDataSource>

@property (weak, nonatomic) IBOutlet UICollectionView *collectionView;

@end


@implementation OCTMainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
    [self.collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:kReuseCellID];
    self.title = [self.collectionView.collectionViewLayout description];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return 13;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:kReuseCellID forIndexPath:indexPath];
    cell.contentView.backgroundColor = [UIColor greenColor];
    
    return cell;
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    [self.collectionView.collectionViewLayout invalidateLayout];
}

@end

File: Swift/OCTCustomCollectionViewLayout/OCTCustomCollectionViewLayout/OCTMainViewController.swift
//
//  ViewController.swift
//  OCTCustomCollectionViewLayout
//
//  Created by dmitry.brovkin on 3/31/17.
//  Copyright © 2017 dmitry.brovkin. All rights reserved.
//

import UIKit

private let kReuseCellID = "reuseCellID"
private let kImgName = "img_"
private let kTotalImgs = 10

class OCTMainViewController: UIViewController, UICollectionViewDataSource {
    @IBOutlet private weak var collectionView: UICollectionView!

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.automaticallyAdjustsScrollViewInsets = false
        self.title = collectionView.collectionViewLayout.description
        
        self.collectionView.contentInset = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
        self.collectionView.register(UINib(nibName: "\(OCTCollectionViewCell.self)", bundle: nil), forCellWithReuseIdentifier: kReuseCellID)
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 100
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let imgNumber = indexPath.item % kTotalImgs + 1
        let imageName = kImgName + "\(imgNumber)"
        
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: kReuseCellID, for: indexPath) as! OCTCollectionViewCell
        cell.imgView!.image = UIImage(named: imageName)
        
        return cell
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)
        self.collectionView.collectionViewLayout.invalidateLayout()
    }
}

